### ISSUE TITLE
Gradient accumulation using DistributedDataParallel & AMP

### ISSUE THREAD

**Issue by julianzaidi** [NVIDIA/apex#237]
Thanks to `amp.scale_loss(loss, optimizer, delay_unscale=True)`, `apex.amp` provides a nice solution when user wants to perform gradient accumulation across iterations.

With `delay_unscale=True`, scaling of the loss will be the same during accumulation, which allows the optimizer to perform its `step()` operation, if gradient is not overflowing. However, is it the same behaviour when performing multi-GPU training using `DistributedDataParallel` ?

In other words, let's say I want to train a model on a machine that has 4 GPUs. If I perform gradient accumulations with `apex.amp` and `delay_unscale=True` on each GPU, will the scaling of the loss be the same on `GPU0`, `GPU1`, `GPU3` and `GPU4` ?

This question can be extended when performing training on more than 1 machine.

**Comment by mcarilli** [comment]
With `torch.nn.parallel.DistributedDataParallel` or `apex.parallel.DistributedDataParallel`, the gradient allreduce across processes occurs before any inf/nan checking, so by the time gradients are checked for infs/nans within each process, these gradients are bitwise identical across processes.  Therefore, each process will find the same infs/nans, update its loss scale the same way as all the others, and the loss scale should remain symmetric across all machines.  If you do NOT observe that loss scaling and step skipping is symmetric across processes, that's a bug and I would definitely like to know about it.

**Comment by mcarilli** [comment]
By the way, I'm about to merge a PR that makes delay_unscale=True unnecessary.  delay_unscale will still be supported, but as an optional performance optimization rather than a necessity.  If your code currently works with delay_unscale=True you don't need to change it.
